sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +58,81 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val arrayClass = classOf[GenericArrayData].getName
-    val values = ctx.freshName("values")
-    ctx.addMutableState("Object[]", values, s"this.$values = null;")
-
-    ev.copy(code = s"""
-      this.$values = new Object[${children.size}];""" +
-      ctx.splitExpressions(
-        ctx.INPUT_ROW,
-        children.zipWithIndex.map { case (e, i) =>
-          val eval = e.genCode(ctx)
-          eval.code + s"""
-            if (${eval.isNull}) {
-              $values[$i] = null;
-            } else {
-              $values[$i] = ${eval.value};
-            }
-           """
-        }) +
-      s"""
-        final ArrayData ${ev.value} = new $arrayClass($values);
-        this.$values = null;
-      """, isNull = "false")
+    val array = ctx.freshName("array")
+
+    val et = dataType.elementType
+    val evals = children.map(e => e.genCode(ctx))
+    val isPrimitiveArray = ctx.isPrimitiveType(et)
+    val (preprocess, arrayData) =
+      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
+
+    val assigns = if (isPrimitiveArray) {
+      val primitiveTypeName = ctx.primitiveTypeName(et)
+      evals.zipWithIndex.map { case (eval, i) =>
+        eval.code + s"""
+         if (${eval.isNull}) {
+           $arrayData.setNullAt($i);
+         } else {
+           $arrayData.set$primitiveTypeName($i, ${eval.value});
+         }
+       """
+      }
+    } else {
+      evals.zipWithIndex.map { case (eval, i) =>
+        eval.code + s"""
+         if (${eval.isNull}) {
+           $array[$i] = null;
+         } else {
+           $array[$i] = ${eval.value};
+         }
+       """
+      }
+    }
+    ev.copy(code =
+      preprocess +
+      ctx.splitExpressions(ctx.INPUT_ROW, assigns) +
+      s"\nfinal ArrayData ${ev.value} = $arrayData;\n",
+      isNull = "false")
   }
 
   override def prettyName: String = "array"
 }
 
+private [sql] object GenArrayData {
+  // This function returns Java code pieces based on DataType and isPrimitive
+  // for allocation of ArrayData class
+  def getCodeArrayData(
+      ctx: CodegenContext,
+      dt: DataType,
+      size: Int,
+      isPrimitive : Boolean,
+      array: String): (String, String) = {
+    if (!isPrimitive) {
+      val arrayClass = classOf[GenericArrayData].getName
+      ctx.addMutableState("Object[]", array,
+        s"this.$array = new Object[${size}];")
+      ("", s"new $arrayClass($array)")
+    } else {
+      val baseArray = ctx.freshName("baseArray")
```

Comment:
this is counter-intuitive, the `array` parameter should be the name of the underlying array, but here you create another name for the underlying array, and use `array` as the name of the result `ArrayData`